File: backend/app/services/ocr_space.py
from __future__ import annotations

from typing import Any

import httpx

from app.core.config import get_settings
from app.core.errors import ServiceError

OCR_SPACE_URL = "https://api.ocr.space/parse/image"


def _normalize_error(value: Any) -> str:
    if isinstance(value, str):
        return value.strip()
    if isinstance(value, list):
        return " ".join(str(item).strip() for item in value if str(item).strip())
    return ""


async def extract_text_with_ocr_space(
    *, file_bytes: bytes, file_name: str, mime_type: str
) -> str:
    settings = get_settings()
    if not settings.ocr_space_api_key:
        raise ServiceError("Receipt OCR service is not configured on this server.", 500)

    files = {"file": (file_name, file_bytes, mime_type)}
    data = {
        "isOverlayRequired": "false",
        "OCREngine": "2",
        "language": "eng",
        "scale": "true",
    }
    headers = {"apikey": settings.ocr_space_api_key}

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                OCR_SPACE_URL,
                data=data,
                files=files,
                headers=headers,
            )
    except httpx.RequestError as exc:
        raise ServiceError("Receipt OCR service is unavailable right now. Please try again.") from exc

    payload: Any
    try:
        payload = response.json()
    except ValueError:
        payload = {}

    if response.status_code >= 400:
        raise ServiceError(
            f"Receipt OCR request failed with status {response.status_code}."
        )

    is_errored = bool(payload.get("IsErroredOnProcessing")) if isinstance(payload, dict) else True
    exit_code = payload.get("OCRExitCode") if isinstance(payload, dict) else None
    if is_errored or exit_code != 1:
        error_message = _normalize_error(payload.get("ErrorMessage") if isinstance(payload, dict) else "")
        details = (
            str(payload.get("ErrorDetails", "")).strip() if isinstance(payload, dict) else ""
        )
        message = " ".join(part for part in [error_message, details] if part).strip()
        raise ServiceError(
            message or "Could not read this receipt image. Try a clearer, front-facing photo."
        )

    parsed_results = payload.get("ParsedResults") if isinstance(payload, dict) else None
    parsed_lines: list[str] = []
    if isinstance(parsed_results, list):
        for result in parsed_results:
            if isinstance(result, dict):
                text = result.get("ParsedText")
                if isinstance(text, str) and text.strip():
                    parsed_lines.append(text.strip())

    parsed_text = "\n".join(parsed_lines).strip()
    if not parsed_text:
        raise ServiceError("No readable text was found on this receipt image.", 422)

    return parsed_text

File: lib/ocr/ocrSpace.ts
import "server-only";

const OCR_SPACE_API_URL = "https://api.ocr.space/parse/image";

type OcrSpaceParsedResult = {
  ParsedText?: string;
  ErrorMessage?: string | string[];
};

type OcrSpaceResponse = {
  OCRExitCode?: number;
  IsErroredOnProcessing?: boolean;
  ErrorMessage?: string | string[];
  ErrorDetails?: string;
  ParsedResults?: OcrSpaceParsedResult[];
};

type ExtractTextInput = {
  fileBytes: ArrayBuffer;
  fileName: string;
  mimeType: string;
};

function getRequiredOcrSpaceKey() {
  const value = process.env.OCR_SPACE_API_KEY?.trim();
  if (!value) {
    throw new Error("The server is missing OCR_SPACE_API_KEY.");
  }
  return value;
}

function normalizeErrorList(value: string | string[] | undefined) {
  if (!value) return "";
  return Array.isArray(value) ? value.filter(Boolean).join(" ") : value;
}

export async function extractTextWithOcrSpace({
  fileBytes,
  fileName,
  mimeType,
}: ExtractTextInput) {
  const formData = new FormData();
  formData.set("isOverlayRequired", "false");
  formData.set("OCREngine", "2");
  formData.set("language", "eng");
  formData.set("scale", "true");
  formData.set("file", new Blob([fileBytes], { type: mimeType }), fileName);

  const response = await fetch(OCR_SPACE_API_URL, {
    method: "POST",
    headers: {
      apikey: getRequiredOcrSpaceKey(),
    },
    body: formData,
  });

  if (!response.ok) {
    throw new Error("OCR service is unavailable right now. Please try again shortly.");
  }

  const payload = (await response.json().catch(() => ({}))) as OcrSpaceResponse;
  const apiError = normalizeErrorList(payload.ErrorMessage);
  const details = payload.ErrorDetails?.trim() ?? "";

  if (payload.IsErroredOnProcessing || payload.OCRExitCode !== 1) {
    const message = [apiError, details].filter(Boolean).join(" ").trim();
    throw new Error(
      message || "Could not read this receipt image. Try a clearer, front-facing photo."
    );
  }

  const parsedText = (payload.ParsedResults ?? [])
    .map((result) => result.ParsedText?.trim() ?? "")
    .filter((text) => text.length > 0)
    .join("\n")
    .trim();

  if (!parsedText) {
    throw new Error("No readable text was found on this receipt image.");
  }

  return parsedText;
}
